R/add_hri_variables.R

```diff
@@ -0,0 +1,137 @@
+#' \item{keep_flag = 3 when the gpprac is not English and the postcode is missing}
+#' \item{keep_flag = 4 when the gpprac is not English and the postcode is a dummy}
+#' }
+#' The intention is to only keep the records where keep_flag = 0
+#'
+#' @param data A data frame, usually an SLF Individual File
+#' @param slf_pc_lookup_path The path to the Source postcode lookup, defaults
+#' to [get_slf_postcode_path()]
+#'
+#' @return A data frame with the variable 'keep_flag'
+flag_non_scottish_residents <- function(data,
+                                        slf_pc_lookup_path = get_slf_postcode_path()) {
+  check_variables_exist(data, c("postcode", "gpprac"))
+
+  # Make a lookup of postcode areas, which consist of the first characters
+  # of the postcode
+  pc_areas <- arrow::read_parquet(slf_pc_lookup_path) %>%
+    dplyr::mutate(
+      pc_area = stringr::str_match(postcode, "^[A-Z]{1,3}"),
+      scot_flag = 0L
+    ) %>%
+    dplyr::distinct(pc_area, scot_flag)
+
+  # Create a flag, 'keep_flag', to determine whether individuals are Scottish
+  # residents or not
+  return_data <- data %>%
+    dplyr::mutate(pc_area = stringr::str_match(postcode, "^[A-Z]{1,3}")) %>%
+    dplyr::left_join(pc_areas, by = "pc_area") %>%
+    dplyr::mutate(
+      dummy_postcode = .data$postcode %in% c("BF010AA", "NF1 1AB", "NK010AA") |
+        stringr::str_sub(.data$postcode, 1, 4) %in% c("ZZ01", "ZZ61"),
+      eng_prac = .data$gpprac %in% c(99942, 99957, 99961, 99976, 99981, 99995, 99999),
+      keep_flag = dplyr::case_when(
+        .data$scot_flag == 0L ~ 0L,
+        (is_missing(.data$postcode) | .data$dummy_postcode) & is.na(.data$gpprac) ~ 2L,
+        !.data$eng_prac & is_missing(.data$postcode) ~ 3L,
+        !.data$eng_prac & .data$dummy_postcode ~ 4L,
+        TRUE ~ 1L
+      )
+    ) %>%
+    dplyr::select(-"dummy_postcode", -"eng_prac")
+
+  return(return_data)
+}
+
+#' Add HRI variables to an SLF Individual File
+#'
+#' @details Filters the dataset to only include Scottish residents, then
+#' creates a lookup where HRIs are calculated at Scotland, Health Board, and
+#' LCA level. Then joins on this lookup by chi/anon_chi.
+#'
+#' @param data An SLF individual file
+#' @param chi_variable Defaults to "anon_chi"
+#'
+#' @return The individual file with HRI variables matched on
+#' @export
+add_hri_variables <- function(data,
+                              chi_variable = "anon_chi") {
+  hri_lookup <- data %>%
+    dplyr::select(
+      "year",
+      chi_variable,
+      "postcode",
+      "gpprac",
+      "lca",
+      "hbrescode",
+      "health_net_cost",
+      "acute_episodes",
+      "mat_episodes",
+      "mh_episodes",
+      "gls_episodes",
+      "op_newcons_attendances",
+      # op_newcons_dnas,
+      "ae_attendances",
+      "pis_paid_items",
+      "ooh_cases"
+    ) %>%
+    flag_non_scottish_residents() %>%
+    dplyr::filter(keep_flag == 0L) %>%
```

Comment:
I'm a bit confused as to why we do all the above, to then only keep people with a Scottish postcode.

It probably makes the most sense for `keep_flag` to be called `scottish_resident` and be T/F. Then we can do `dplyr::filter(scottish_resident)` and the code will read nicely.